File: package/main/src/Error/safeExecute.ts
/**
 * Represents a successful execution result containing a value
 * @template V The type of the success value
 */
interface SuccessType<V> {
  type: "success";
  value: V;
}

/**
 * Represents an error result containing an error object
 * @template E The type of the error
 */
interface ErrorType<E> {
  type: "error";
  error: E;
}

/**
 * Union type representing either a successful result or an error result
 * @template V The type of the success value
 * @template E The type of the error
 */
export type Result<V, E> = SuccessType<V> | ErrorType<E>;
/**
 * Creates an error result
 * @template E The type of the error
 * @param error The error object
 * @returns An ErrorType containing the error
 */
const errorFunction = <E>(error: E): ErrorType<E> => ({ type: "error", error });
/**
 * Creates a success result
 * @template V The type of the success value
 * @param value The success value
 * @returns A SuccessType containing the value
 */
const successFunction = <V>(value: V): SuccessType<V> => ({
  type: "success",
  value,
});
/**
 * Safely executes a callback function and returns a Result type
 * Catches any errors and wraps them in a Result type
 * @template V The type of the success value
 * @template E The type of the error (defaults to Error)
 * @param callback The function to execute safely
 * @returns A Result containing either the successful value or the caught error
 */
export const safeExecute = <V, E = Error>(callback: () => V): Result<V, E> => {
  try {
    return successFunction<V>(callback());
  } catch (error) {
    return errorFunction<E>(error as E);
  }
};

File: package/umt_python/src/error/safe_execute.py
from collections.abc import Callable
from dataclasses import dataclass
from typing import Generic, TypeVar

V = TypeVar("V")
E = TypeVar("E", bound=Exception)


@dataclass
class Success(Generic[V]):
    """Represents a successful execution result containing a value."""

    type: str = "success"
    value: V = None  # type: ignore
    error: None = None


@dataclass
class Error(Generic[E]):
    """Represents an error result containing an error object."""

    type: str = "error"
    error: E = None  # type: ignore
    value: None = None


Result = Success[V] | Error[E]


def safe_execute(callback: Callable[[], V]) -> Result[V, Exception]:
    """
    Safely executes a callback function and returns a Result type.
    Catches any errors and wraps them in a Result type.

    Args:
        callback: The function to execute safely

    Returns:
        A Result containing either the successful value or the caught error

    Example:
        >>> result = safe_execute(lambda: 1 / 0)
        >>> result.type
        'error'
        >>> result = safe_execute(lambda: 1 + 1)
        >>> result.value
        2
    """
    try:
        value = callback()
        return Success(type="success", value=value)
    except Exception as error:
        return Error(type="error", error=error)
